File: javascript/scale-generator/scale-generator.js
const sharps = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B'];
const flats = ['F', 'Gb', 'G', 'Ab', 'A', 'Bb', 'B', 'C', 'Db', 'D', 'Eb', 'E'];
const usesFlats = ['F', 'Bb', 'Eb', 'Ab', 'Db', 'Gb', 'd', 'g', 'c', 'f', 'bb', 'eb'];
const steps = {
  "m": 1,
  "M": 2,
  "A": 3,
}

export class Scale {
  constructor(tonic) {
    this.tonic = tonic;
  }

  chromatic() {
    const scale = usesFlats.includes(this.tonic) ? flats : sharps;
    let scalePos = scale.indexOf(this.tonic)
    return scale.map((_, i) => scale[(scalePos + i) % scale.length])
  }

  interval(intervals) {
    const currTonic = this.tonic[0].toUpperCase() + this.tonic.slice(1);
    const scale = usesFlats.includes(this.tonic) ? flats : sharps;
    let scalePos = scale.indexOf(currTonic);

    return [...intervals].map((step) => {
      const nextNote = scale[scalePos]
      scalePos = (scalePos + steps[step]) % scale.length;
      return nextNote
    });
  }
}

File: csharp/scale-generator/ScaleGenerator.cs
using System;
using System.Collections.Generic;

public static class ScaleGenerator
{
    static List<string> sharps = new List<string> { "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B" };
    static List<string> flats = new List<string> { "F", "Gb", "G", "Ab", "A", "Bb", "B", "C", "Db", "D", "Eb", "E" };
    static List<string> usesFlats = new List<string> { "F", "Bb", "Eb", "Ab", "Db", "Gb", "d", "g", "c", "f", "bb", "eb" };

    private static readonly Dictionary<Char, int> stepMap = new Dictionary<Char, int>() {
        { 'm', 1 },
        { 'M', 2 },
        { 'A', 3 },
    };

    public static string[] Chromatic(string tonic)
    {
        return tonic == "C" ? sharps.ToArray() : flats.ToArray();
    }

    public static string[] Interval(string tonic, string pattern)
    {
        var res = new List<String>();
        List<string> notes = usesFlats.Contains(tonic) ? flats : sharps;
        var index = notes.IndexOf(tonic.ToUpper()[0] + tonic.Substring(1));

        foreach (var step in pattern) {
            res.Add(notes[index]);
            index = (index + stepMap[step]) % notes.Count;
        }
        return res.ToArray();
    }
}
